File: src/main/java/com/extendedae_plus/common/impl/menuLocator/WirelessTerminalLocator.java
package com.extendedae_plus.common.impl.menuLocator;

import appeng.api.networking.IGrid;
import appeng.items.tools.powered.WirelessTerminalItem;
import appeng.menu.locator.ItemMenuHostLocator;
import appeng.menu.locator.MenuLocators;
import com.extendedae_plus.integration.ContextModLoaded;
import de.mari_023.ae2wtlib.api.terminal.ItemWT;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import top.theillusivec4.curios.api.CuriosApi;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

public final class WirelessTerminalLocator {
    public static Optional<TerminalInfo> locate(Player player) {
        var inventory = player.getInventory();

        if (ContextModLoaded.curios.isLoaded()) {
            try {
                AtomicReference<TerminalInfo> info = new AtomicReference<>();
                CuriosApi.getCuriosInventory(player).ifPresent(handler -> {
                    var resultOptional = handler.findFirstCurio(
                            itemStack -> itemStack.getItem() instanceof WirelessTerminalItem);
                    if (resultOptional.isEmpty()) return;
                    var result = resultOptional.get();

                    info.set(new TerminalInfo(result.stack(), (WirelessTerminalItem) result.stack().getItem(),
                            new LocatedSlotContext(player, false, -1,
                                    result.slotContext().identifier(), result.slotContext().index())));
                });
                if (info.get() != null) return Optional.of(info.get());
            } catch (Throwable ignore) {
            }
        }

        if (inventory.offhand.getFirst().getItem() instanceof WirelessTerminalItem terminal)
            return Optional.of(new TerminalInfo(inventory.offhand.getFirst(), terminal,
                    new LocatedSlotContext(player, true, -1, "", -1)));

        for (int invIndex = 0; invIndex < inventory.items.size(); invIndex++) {
            var item = inventory.items.get(invIndex);
            if (!(item.getItem() instanceof WirelessTerminalItem terminal))
                continue;

            return Optional.of(new TerminalInfo(item, terminal,
                    new LocatedSlotContext(player, false, invIndex, "", -1)));
        }

        return Optional.empty();
    }

    public record TerminalInfo(ItemStack terminalStack, WirelessTerminalItem terminal,
                               LocatedSlotContext context) {
        public boolean isWTLibTerminal() {
            if (ContextModLoaded.ae2wtlib.isLoaded())
                return terminalStack.getItem() instanceof ItemWT;
            else return false;
        }

        public Optional<ItemMenuHostLocator> getMenuLocator() {
            var curioType = context.curioType;
            var curioIndex = context.curioIndex;
            var slot = context.invIndex;

            ItemMenuHostLocator locator = null;
            if (curioIndex >= 0) locator = new CuriosItemLocator(curioType, curioIndex, null);
            else if (context.offhand) locator = MenuLocators.forHand(context.player, InteractionHand.OFF_HAND);
            else if (slot >= 0) locator = MenuLocators.forInventorySlot(slot);

            return Optional.ofNullable(locator);
        }

        public Optional<IGrid> grid() {
            var menuLocator = getMenuLocator();
            if (isWTLibTerminal() && menuLocator.isPresent()) {
                var node = ((ItemWT) terminal)
                        .getMenuHost(context.player, menuLocator.get(), null)
                        .getActionableNode();
                if (node == null) return Optional.empty();
                return Optional.ofNullable(node.getGrid());
            }

            return Optional.ofNullable(
                    terminal.getLinkedGrid(terminalStack, context.player.level(), null));
        }
    }

    public record LocatedSlotContext(Player player, boolean offhand, int invIndex, String curioType, int curioIndex) {
    }
}

File: src/main/kotlin/com/extendedae_plus/common/impl/WirelessTerminalLocator.kt
package com.extendedae_plus.common.impl

import appeng.api.networking.IGrid
import appeng.items.tools.powered.WirelessTerminalItem
import appeng.menu.locator.ItemMenuHostLocator
import appeng.menu.locator.MenuLocators
import com.extendedae_plus.common.impl.menuLocator.CuriosItemLocator
import com.extendedae_plus.integration.helper.ContextModLoaded
import com.fish.fishlib.util.extension.cast
import de.mari_023.ae2wtlib.api.terminal.ItemWT
import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import top.theillusivec4.curios.api.CuriosApi
import kotlin.jvm.optionals.getOrNull

object WirelessTerminalLocator {
    fun locate(player: Player): InfoTerminal? {
        val inventory = player.getInventory()

        if (ContextModLoaded.Curios())
            CuriosApi.getCuriosInventory(player)
                .getOrNull()
                ?.let {
                    val result = it
                        .findFirstCurio { it.item is WirelessTerminalItem }
                        .getOrNull()
                        ?: return@let null
                    InfoTerminal(
                        result.stack(),
                        result.stack().item.cast(),
                        ContextSlotLocated(
                            player,
                            false,
                            -1,
                            result.slotContext().identifier(),
                            result.slotContext().index()
                        )
                    )
                }?.let { return it }

        (inventory.offhand[0].item as? WirelessTerminalItem)?.let {
            return InfoTerminal(
                inventory.offhand[0],
                it,
                ContextSlotLocated(
                    player,
                    true,
                    -1,
                    "",
                    -1
                )
            )
        }

        for (indexInv in inventory.items.indices) {
            val item = inventory.items[indexInv]
            (item.item as? WirelessTerminalItem)?.let {
                return InfoTerminal(
                    item,
                    it,
                    ContextSlotLocated(
                        player,
                        false,
                        indexInv,
                        "",
                        -1
                    )
                )
            }
        }

        return null
    }

    @JvmRecord
    data class InfoTerminal(
        val terminalStack: ItemStack,
        val terminal: WirelessTerminalItem,
        val context: ContextSlotLocated
    ) {
        val isWTLibTerminal: Boolean
            get() {
                return if (ContextModLoaded.AE2wtlib()) terminalStack.item is ItemWT
                else false
            }

        val menuLocator: ItemMenuHostLocator?
            get() {
                val curioType = context.curioType
                val curioIndex = context.curioIndex
                val slot = context.invIndex

                return when {
                    curioIndex >= 0 -> CuriosItemLocator(curioType, curioIndex, null)
                    context.offhand -> MenuLocators.forHand(context.player, InteractionHand.OFF_HAND)
                    slot >= 0 -> MenuLocators.forInventorySlot(slot)
                    else -> null
                }
            }

        fun grid(): IGrid? {
            val menuLocator = this.menuLocator

            return if (this.isWTLibTerminal && menuLocator != null) {
                (terminal as ItemWT)
                    .getMenuHost(context.player, menuLocator, null)!!
                    .getActionableNode()
                    ?.grid
            } else terminal.getLinkedGrid(
                terminalStack,
                context.player.level(),
                null
            )
        }
    }

    @JvmRecord
    data class ContextSlotLocated(
        val player: Player,
        val offhand: Boolean,
        val invIndex: Int,
        val curioType: String,
        val curioIndex: Int
    )
}
